django/db/backends/postgresql/schema.py

```diff
@@ -41,11 +43,14 @@ def _field_indexes_sql(self, model, field):
 
     def _field_data_type(self, field):
         if field.is_relation:
-            return field.rel_db_type(self.connection)
-        return self.connection.data_types.get(
-            field.get_internal_type(),
-            field.db_type(self.connection),
-        )
+            data_type = field.rel_db_type(self.connection)
+        else:
+            data_type = self.connection.data_types.get(
+                field.get_internal_type(),
+                field.db_type(self.connection),
+            )
+        # Remove everything inside parentheses and brackets
+        return re.sub(r'(\[|\()[^()]*(\)|\])', '', data_type)
```

Comment:
I would like to avoid using of regexp here, it is always clunky. We should take into account that only the `ArrayField` is affected but it's more complicated then it looks like, e.g. if we decrease a size and change a `base_field` in a single alteration then we still have a data loss:
- `_field_data_type(ArrayField(IntegerField(), size=5))` returns `integer`,
- `_field_data_type(ArrayField(CharField(max_length=16), size=4` returns `varchar`.

So for `ArrayField` we should cast only when `base_field` is different and `size` (in any dimension) is not smaller than the previous one. It's complicated :disappointed: and would require a very special treatment of `ArrayField`'s. (\cc @charettes )